Add xtensa esp32 asm! tests

KerryRJ opened the pull request:
According to this link [https://rustc-dev-guide.rust-lang.org/tests/running.html#running-tests-manually](url) I can run tests manually with this command 

> rustc +esp src/test/assembly/asm/xtensa-types.rs --target=xtensa-esp32-none-elf

I am not sure if I am doing something wrong but the command just runs and doesn't print out anything. I am assuming that the tests are running through successfully.

I haven't found a way to get these tests to run against the xtensa-esp32-none-elf target using

> ./x.py test --stage 2 src/test/assembly/asm/xtensa-types.rs

The F64 tests are currently commented Todos as I have not come across any documentation that I can use to write them.

MabezDev commented:
Thanks! Would you mind switching the target branch of this PR to `xtensa-asm-impl`?

> The F64 tests are currently commented Todos as I have not come across any documentation that I can use to write them.

I can't find anything either,  grepping esp-idf shows no usages of the dp register - I don't think this register is something that should be managed in inline asm.

KerryRJ commented:
Must I remove the breg test?

MabezDev commented:
@KerryRJ would you mind rebasing to fix conflicts, and then remove the bool register tests? I think we're ready to merge this :)

MabezDev commented:
Perfect, thank you!

KerryRJ commented:
Rebased, removed breg tests and F64 comments.

I have been working on the updates to xtensa-lx based on these changes but I'm getting a panic when running alloc, mem, and ram esp32-hal examples. 

It is panicking on the last line in this sample
```Rust
    fn with_locked_heap<R>(&self, f: impl FnOnce(&mut Heap) -> R) -> R {
        (&self.heap).lock(|heap| match heap {
            None => {
                let start = (self.start_addr)() as usize;
```
I get a panic when I try and assign _text_heap_start to a temp variable from here

```Rust
extern "C" {
    static _heap_start: u8;
    static _heap_end: u8;
    static _text_heap_start: u8;
    static _text_heap_end: u8;
    static _external_heap_start: u8;
    static _external_heap_end: u8;
}
```
It appears that it has something to do with the memory.x file not generating the _text_heap_start variable (I don't yet know how this file is used to generate code). I am also unsure as to whether it is the asm! changes or my adjustments to xtensa-lx.

MabezDev commented:
> Rebased, removed breg tests and F64 comments.
> 
> I have been working on the updates to xtensa-lx based on these changes but I'm getting a panic when running alloc, mem, and ram esp32-hal examples.
> 
> It is panicking on the last line in this sample
> 
> ```rust
>     fn with_locked_heap<R>(&self, f: impl FnOnce(&mut Heap) -> R) -> R {
>         (&self.heap).lock(|heap| match heap {
>             None => {
>                 let start = (self.start_addr)() as usize;
> ```
> 
> I get a panic when I try and assign _text_heap_start to a temp variable from here
> 
> ```rust
> extern "C" {
>     static _heap_start: u8;
>     static _heap_end: u8;
>     static _text_heap_start: u8;
>     static _text_heap_end: u8;
>     static _external_heap_start: u8;
>     static _external_heap_end: u8;
> }
> ```
> 
> It appears that it has something to do with the memory.x file not generating the _text_heap_start variable (I don't yet know how this file is used to generate code). I am also unsure as to whether it is the asm! changes or my adjustments to xtensa-lx.

Hard to know without seeing the code, mind PRing those and we can work through it? Marking them as draft should be okay :)

MabezDev commented:
FYI I've just pushed xtensa-lx & lx-rt releases with the new asm. I didn't run into the issues you had when using the new crates, so perhaps try the released version and see if you still have issues?

KerryRJ commented:
I have just rebuilt llvm and rust from scratch. Just tested the lx and lx-rt new releases and still getting a LoadProhibited exception.

Stepping through the code points to the _text_heap_start value from memory.x not pulling through. I will rest for now and tackle this again tomorrow.
